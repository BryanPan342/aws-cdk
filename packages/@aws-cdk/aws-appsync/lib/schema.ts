import { IBucket } from '@aws-cdk/aws-s3';
import { AssetSchemaImpl, CodeSchemaImpl, SchemaImpl, S3SchemaImpl } from './private/schema-implementation';

/**
 * An interface for S3 Location for Schema Implementation
 */
export interface S3Location {
  /**
   * The S3 bucket that stores the GraphQL Schema
   */
  readonly bucket: IBucket;

  /**
   * The object key for the GraphQL Schema
   */
  readonly key: string;
}

/**
 * The Schema for a GraphQL Api
 *
 * If no options are configured, schema will be generated
 * code-first.
 */
export abstract class Schema extends SchemaImpl {
  /**
   * Generate a Schema from file
   *
   * @returns `Schema` with immutable schema defintion
   * @param filePath the file path of the schema file
   */
  public static fromAsset(filePath: string): Schema {
    return new AssetSchemaImpl(filePath);
  }

  /**
   * Point GraphQL Schema to a S3 Location
   *
   * @returns `Schema` with immutable schema defintion
   * @param s3Location the s3 location of the schema
   */
  public static fromS3Location(s3Location: S3Location): Schema {
    return new S3SchemaImpl(s3Location);
  }

  /**
   * Generate a Schema that can be edited through a code-first
   * approach
   *
   * @returns `Schema` with mutable schema defintion
   */
  public static fromCode(): Schema {
    return new CodeSchemaImpl();
  }
}