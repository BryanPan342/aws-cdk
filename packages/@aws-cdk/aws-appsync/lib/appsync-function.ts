import { Construct, Resource } from '@aws-cdk/core';
import { CfnFunctionConfiguration } from './appsync.generated';
import { BaseDataSource } from './data-source';
import { IGraphqlApi } from './graphqlapi-base';
import { MappingTemplate } from './mapping-template';

/**
 * the base properties for AppSync Functions
 */
export interface BaseAppsyncFunctionProps {
  /**
   * the name of the AppSync Function
   */
  readonly name: string;
  /**
   * the function version of this AppSync Function
   *
   * @default '1.0.0'
   */
  readonly functionVersion?: string;
  /**
   * the description for this AppSync Function
   *
   * @default - no description
   */
  readonly description?: string;
  /**
   * the request mapping template for the AppSync Function
   *
   * @default - no request mapping template
   */
  readonly requestMappingTemplate?: MappingTemplate;
  /**
   * the response mapping template for the AppSync Function
   *
   * @default - no response mapping template
   */
  readonly responseMappingTemplate?: MappingTemplate;
}

/**
 * the CDK properties for AppSync Functions
 */
export interface AppsyncFunctionProps extends BaseAppsyncFunctionProps {
  /**
   * the GraphQL Api linked to this AppSync Function
   */
  readonly api: IGraphqlApi;
  /**
   * the data source linked to this AppSync Function
   */
  readonly dataSource: BaseDataSource;
}

/**
 * AppSync Functions are local functions that perform certain operations
 * onto a backend data source. Developers can compose operations (Functions)
 * and execute them in sequence with Pipeline Resolvers.
 *
 * @resource AWS::AppSync::FunctionConfiguration
 */
export class AppsyncFunction extends Resource {
  /**
   * the name of this AppSync Function
   *
   * @attribute Name
   */
  public readonly functionName: string;
  /**
   * the ARN of the AppSync function
   *
   * @attribute
   */
  public readonly functionArn: string;
  /**
   * the ID of the AppSync function
   *
   * @attribute
   */
  public readonly functionId: string;
  /**
   * the data source of this AppSync Function
   *
   * @attribute DataSourceName
   */
  public readonly dataSource: BaseDataSource;

  private readonly function: CfnFunctionConfiguration;

  public constructor(scope: Construct, id: string, props: AppsyncFunctionProps) {
    super(scope, id);
    this.function = new CfnFunctionConfiguration(this, 'Resource', {
      name: props.name,
      description: props.description,
      apiId: props.api.apiId,
      dataSourceName: props.dataSource.name,
      functionVersion: props.functionVersion ?? '1.0.0',
      requestMappingTemplate: props.requestMappingTemplate?.renderTemplate(),
      responseMappingTemplate: props.responseMappingTemplate?.renderTemplate(),
    });
    this.functionName = this.function.attrName;
    this.functionArn = this.function.attrFunctionArn;
    this.functionId = this.function.attrFunctionId;
    this.dataSource = props.dataSource;

    props.api.addSchemaDependency(this.function);
  }
}