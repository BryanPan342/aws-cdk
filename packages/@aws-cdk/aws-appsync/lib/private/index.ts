export * from './mapping-template';
export * from './schema-implementation';
export * from './schema';