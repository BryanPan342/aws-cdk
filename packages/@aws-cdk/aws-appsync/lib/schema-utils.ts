import { IBucket } from '@aws-cdk/aws-s3';

/**
 * An interface for S3 Location for Schema Implementation
 */
export interface S3Location {
  /**
   * The S3 bucket that stores the GraphQL Schema
   */
  readonly bucket: IBucket;

  /**
   * The object key for the GraphQL Schema
   */
  readonly key: string;
}