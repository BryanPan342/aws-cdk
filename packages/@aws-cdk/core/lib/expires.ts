import { Duration } from './duration';
/**
 * Represents a date of expiration.
 *
 * The amount can be specified either as a Date object, timestamp, Duration or string.
 */
export class Expires {
  /**
   * Expire at the specified date
   * @param d date to expire at
   */
  public static atDate(d: Date) { return new Expires(d.toUTCString(), d); }

  /**
   * Expire at the specified timestamp
   * @param t timestamp in unix milliseconds
   */
  public static atTimestamp(t: number) { return Expires.atDate(new Date(t)); }

  /**
   * Expire once the specified duration has passed since deployment time
   * @param t the duration to wait before expiring
   */
  public static after(t: Duration) { return Expires.atDate(new Date(Date.now() + t.toMilliseconds())); }

  /**
   * Expire at specified date, represented as a string
   *
   * @param s the string that represents date to expire at
   */
  public static fromString(s: string) { return new Expires(s, new Date(s)); }

  /**
   * Expiration value as a string
   */
  public readonly value: string;

  /**
   * Expiration value as a Date object
   */
  public readonly date: Date;

  private constructor(value: string, date: Date) {
    this.value = value;
    this.date = date;
  }
}