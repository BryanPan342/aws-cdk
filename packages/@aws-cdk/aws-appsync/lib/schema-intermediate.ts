import { shapeAddition } from './private';
import { Resolver } from './resolver';
import { Directive, IField, IIntermediateType, AddFieldOptions } from './schema-base';
import { BaseTypeOptions, GraphqlType, ResolvableFieldOptions } from './schema-field';

/**
 * Properties for configuring an Intermediate Type
 *
 * @param definition - the variables and types that define this type
 * i.e. { string: GraphqlType, string: GraphqlType }
 *
 * @experimental
 */
export interface IntermediateTypeOptions {
  /**
   * the attributes of this type
   */
  readonly definition: { [key: string]: IField };
}

/**
 * Interface Types are abstract types that includes a certain set of fields
 * that other types must include if they implement the interface.
 *
 * @experimental
 */
export class InterfaceType implements IIntermediateType {
  /**
   * the name of this type
   */
  public readonly name: string;
  /**
   * the attributes of this type
   */
  public readonly definition: { [key: string]: IField };

  public constructor(name: string, props: IntermediateTypeOptions) {
    this.name = name;
    this.definition = props.definition;
  }

  /**
   * Create an GraphQL Type representing this Intermediate Type
   *
   * @param options the options to configure this attribute
   * - isList
   * - isRequired
   * - isRequiredList
   */
  public attribute(options?: BaseTypeOptions): GraphqlType {
    return GraphqlType.intermediate({
      isList: options?.isList,
      isRequired: options?.isRequired,
      isRequiredList: options?.isRequiredList,
      intermediateType: this,
    });
  }

  /**
   * Generate the string of this object type
   */
  public toString(): string {
    return shapeAddition({
      prefix: 'interface',
      name: this.name,
      fields: Object.keys(this.definition).map((key) =>
        `${key}${this.definition[key].argsToString()}: ${this.definition[key].toString()}`),
    });
  }

  /**
   * Add a field to this Interface Type.
   *
   * Interface Types must have both fieldName and field options.
   *
   * @param options the options to add a field
   */
  public addField(options: AddFieldOptions): void {
    if (!options.fieldName || !options.field) {
      throw new Error('Interface Types must have both fieldName and field options.');
    }
    this.definition[options.fieldName] = options.field;
  }
}

/**
 * Properties for configuring an Object Type
 *
 * @param definition - the variables and types that define this type
 * i.e. { string: GraphqlType, string: GraphqlType }
 * @param interfaceTypes - the interfaces that this object type implements
 * @param directives - the directives for this object type
 *
 * @experimental
 */
export interface ObjectTypeOptions extends IntermediateTypeOptions {
  /**
   * The Interface Types this Object Type implements
   *
   * @default - no interface types
   */
  readonly interfaceTypes?: InterfaceType[];
  /**
   * the directives for this object type
   *
   * @default - no directives
   */
  readonly directives?: Directive[];
}

/**
 * Object Types are types declared by you.
 *
 * @experimental
 */
export class ObjectType extends InterfaceType implements IIntermediateType {
  /**
   * The Interface Types this Object Type implements
   *
   * @default - no interface types
   */
  public readonly interfaceTypes?: InterfaceType[];
  /**
   * the directives for this object type
   *
   * @default - no directives
   */
  public readonly directives?: Directive[];
  /**
   * The resolvers linked to this data source
   */
  public resolvers?: Resolver[];

  public constructor(name: string, props: ObjectTypeOptions) {
    const options = {
      definition: props.interfaceTypes?.reduce((def, interfaceType) => {
        return Object.assign({}, def, interfaceType.definition);
      }, props.definition) ?? props.definition,
    };
    super(name, options);
    this.interfaceTypes = props.interfaceTypes;
    this.directives = props.directives;
    this.resolvers = [];

    Object.keys(this.definition).forEach((fieldName) => {
      const field = this.definition[fieldName];
      this.generateResolver(fieldName, field.fieldOptions);
    });
  }

  /**
   * Add a field to this Object Type.
   *
   * Object Types must have both fieldName and field options.
   *
   * @param options the options to add a field
   */
  public addField(options: AddFieldOptions): void {
    if (!options.fieldName || !options.field) {
      throw new Error('Object Types must have both fieldName and field options.');
    }
    this.generateResolver(options.fieldName, options.field.fieldOptions);
    this.definition[options.fieldName] = options.field;
  }

  /**
   * Generate the string of this object type
   */
  public toString(): string {
    return shapeAddition({
      prefix: 'type',
      name: this.name,
      interfaceTypes: this.interfaceTypes,
      directives: this.directives,
      fields: Object.keys(this.definition).map((key) =>
        `${key}${this.definition[key].argsToString()}: ${this.definition[key].toString()}`),
    });
  }

  /**
   * Generate the resolvers linked to this Object Type
   */
  protected generateResolver(fieldName: string, options?: ResolvableFieldOptions): void {
    if (options?.dataSource) {
      if (!this.resolvers) { this.resolvers = []; }
      this.resolvers.push(options.dataSource.createResolver({
        typeName: this.name,
        fieldName: fieldName,
        pipelineConfig: options.pipelineConfig,
        requestMappingTemplate: options.requestMappingTemplate,
        responseMappingTemplate: options.responseMappingTemplate,
      }));
    }
  }
}

/**
 * Input Types are abstract types that define complex objects.
 * They are used in arguments to represent
 *
 * @experimental
 */
export class InputType implements IIntermediateType {
  /**
   * the name of this type
   */
  public readonly name: string;
  /**
   * the attributes of this type
   */
  public readonly definition: { [key: string]: IField };

  public constructor(name: string, props: IntermediateTypeOptions) {
    this.name = name;
    this.definition = props.definition;
  }

  /**
   * Create an GraphQL Type representing this Input Type
   *
   * @param options the options to configure this attribute
   * - isList
   * - isRequired
   * - isRequiredList
   */
  public attribute(options?: BaseTypeOptions): GraphqlType {
    return GraphqlType.intermediate({
      isList: options?.isList,
      isRequired: options?.isRequired,
      isRequiredList: options?.isRequiredList,
      intermediateType: this,
    });
  }

  /**
   * Generate the string of this input type
   */
  public toString(): string {
    return shapeAddition({
      prefix: 'input',
      name: this.name,
      fields: Object.keys(this.definition).map((key) =>
        `${key}${this.definition[key].argsToString()}: ${this.definition[key].toString()}`),
    });
  }

  /**
   * Add a field to this Input Type.
   *
   * Input Types must have both fieldName and field options.
   *
   * @param options the options to add a field
   */
  public addField(options: AddFieldOptions): void {
    if (!options.fieldName || !options.field) {
      throw new Error('Input Types must have both fieldName and field options.');
    }
    this.definition[options.fieldName] = options.field;
  }
}

/**
 * Properties for configuring an Enum Type
 *
 * @param definition - the strings that define this enum type
 *
 * @experimental
 */
export interface EnumTypeOptions {
  /**
   * the attributes of this type
   */
  readonly definition: string[];
}

/**
 * Enum Types are abstract types that includes a set of fields
 * that represent the strings this type can create.
 *
 * @experimental
 */
export class EnumType implements IIntermediateType {
  /**
   * the name of this type
   */
  public readonly name: string;
  /**
   * the attributes of this type
   */
  public readonly definition: { [key: string]: IField };

  public constructor(name: string, props: EnumTypeOptions) {
    this.name = name;
    this.definition = {};
    props.definition.map((fieldName: string) => this.addField({ fieldName }));
  }

  /**
   * Create an GraphQL Type representing this Enum Type
   *
   * @param options the options to configure this attribute
   * - isList
   * - isRequired
   * - isRequiredList
   */
  public attribute(options?: BaseTypeOptions): GraphqlType {
    return GraphqlType.intermediate({
      isList: options?.isList,
      isRequired: options?.isRequired,
      isRequiredList: options?.isRequiredList,
      intermediateType: this,
    });
  }

  /**
   * Generate the string of this enum type
   */
  public toString(): string {
    return shapeAddition({
      prefix: 'enum',
      name: this.name,
      fields: Object.keys(this.definition),
    });
  }

  /**
   * Add a field to this Enum Type
   *
   * To add a field to this Enum Type, you must only configure
   * addField with the fieldName options.
   *
   * @param options the options to add a field
   */
  public addField(options: AddFieldOptions): void {
    if (options.field) {
      throw new Error('Enum Types do not support IField properties. Use the fieldName option.');
    }
    if (!options.fieldName) {
      throw new Error('When adding a field to an Enum Type, you must configure the fieldName option.');
    }
    if (options.fieldName.indexOf(' ') > -1) {
      throw new Error(`The allowed values of an Enum Type must not have white space. Remove the spaces in "${options.fieldName}"`);
    }
    this.definition[options.fieldName] = GraphqlType.string();
  }
}