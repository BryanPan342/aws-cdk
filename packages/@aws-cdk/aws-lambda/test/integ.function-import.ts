import * as iam from '@aws-cdk/aws-iam';
import * as cdk from '@aws-cdk/core';
import * as lambda from '../lib';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'Imports');

const baseFunc = new lambda.Function(stack, 'BaseFunction', {
  code: new lambda.InlineCode('foo'),
  handler: 'index.handler',
  runtime: lambda.Runtime.NODEJS_10_X,
});

const iFunc = lambda.Function.fromFunctionAttributes(stack, 'iFunc', {
  functionArn: baseFunc.functionArn,
});

iFunc.addPermission('iFunc', {
  principal: new iam.ServicePrincipal('cloudformation.amazonaws.com'),
});

const iVersionFunc = lambda.Version.fromVersionAttributes(stack, 'iVersionFunc', {
  version: baseFunc.latestVersion.version,
  lambda: baseFunc,
});

iVersionFunc.addPermission('iVersionFunc', {
  principal: new iam.ServicePrincipal('cloudformation.amazonaws.com'),
});

app.synth();