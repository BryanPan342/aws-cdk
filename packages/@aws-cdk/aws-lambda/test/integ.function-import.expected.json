{
  "Resources": {
    "BaseFunctionServiceRole9D84751B": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "BaseFunction2720903A": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "foo"
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "BaseFunctionServiceRole9D84751B",
            "Arn"
          ]
        },
        "Runtime": "nodejs10.x"
      },
      "DependsOn": [
        "BaseFunctionServiceRole9D84751B"
      ]
    },
    "iFuncA576B2A7": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "BaseFunction2720903A",
            "Arn"
          ]
        },
        "Principal": "cloudformation.amazonaws.com"
      }
    }
  }
}